### PR TITLE
Constify TIMESTAMPTZ OP INTERVAL in constraints

### PR DESCRIPTION
**Pull request by svenklemm**
Constify expressions of the following form in WHERE clause:

column OP timestamptz - interval
column OP timestamptz + interval
column OP interval + timestamptz

Iff interval has no month component.

Since the operators for timestamptz OP interval are marked
as stable they will not be constified during planning.
However, intervals without a month component can be safely
constified during planning as the result of those calculations
do not depend on the timezone setting.

### CODE REVIEW

File: src/plan_expand_hypertable.c [hunk 1]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+ * during planning.
+ */
+static OpExpr *
+constify_timestamptz_op_interval(PlannerInfo *root, OpExpr *constraint)
+{
+	Assert(op->args->length == 2);
+
+	Expr *left = linitial(constraint->args);
+	Expr *right = llast(constraint->args);
+	OpExpr *op;
```
Comment (mkindahl) [inline src/plan_expand_hypertable.c:141]: Nit: would suggest `opexpr` since it actually is an expression, not an operator (i.e., not a function).

File: src/plan_expand_hypertable.c [hunk 2]
```diff
@@ -484,6 +572,16 @@ process_quals(Node *quals, CollectQualCtx *ctx, bool is_outer_join)
...
+			/*
+			 * check for constraints with TIMESTAMPTZ OP INTERVAL calculations
+			 */
+			if ((IsA(left, Var) && IsA(right, OpExpr) &&
+				 castNode(OpExpr, right)->args->length == 2) ||
+				(IsA(left, OpExpr) && IsA(right, Var) && castNode(OpExpr, left)->args->length == 2))
+			{
+				qual = (Expr *) constify_timestamptz_op_interval(ctx->root, op);
+			}
+
```
Comment (mkindahl) [inline src/plan_expand_hypertable.c:584]: Nit: Split this into two separate if statements. There are two distinct cases where the constification should happen and IMHO it is good to make that clear, even if it happen to be identical operations for both cases.

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+
+-- all of these should have all chunk exclusion happening at plan time
+EXPLAIN SELECT time FROM metrics WHERE time < '2000-01-01'::timestamptz - '6h'::interval;
+EXPLAIN SELECT time FROM metrics WHERE time < '2000-01-01'::timestamptz - '1 day'::interval;
+EXPLAIN SELECT time FROM metrics WHERE time < '2000-01-07'::timestamptz - '7 day'::interval;
+-- month calculation are not constified so exclusion happens at execution time
+EXPLAIN SELECT time FROM metrics WHERE time < '2000-02-01'::timestamptz - '1 month'::interval;
+
```
Comment (mkindahl) [inline tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: I'm missing a lot of strange combinations. With the exception of the last case, these are all positives. Can you add tests for the following:

* INTERVAL + TIMESTAMP
* TIMESTAMP - INTERVAL + INTERVAL
* TIMESTAMP + INTERVAL '1 month' - INTERVAL '1 day'
* TIMESTAMP + INTERVAL '1 month' + INTERVAL '1 day'
* INTERVAL '1 month' + TIMESTAMP - INTERVAL '1 day'
* INTERVAL '1 month' + TIMESTAMP + INTERVAL '1 day'
* TIMESTAMP - INTERVAL '1 month - 3 days'
* TIMESTAMP - INTERVAL '1 month - 3 days'

File: src/plan_expand_hypertable.c [hunk 4]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+		constified =
+			DirectFunctionCall2(timestamptz_mi_interval, c_ts->constvalue, c_int->constvalue);
+
+	c_ts = copyObject(c_ts);
+	c_ts->constvalue = constified;
+
+	if (var_on_left)
+		right = (Expr *) c_ts;
+	else
+		left = (Expr *) c_ts;
```
Comment (mkindahl) [inline src/plan_expand_hypertable.c:195]: I don't think you need to make this distinction (of left and right part of expression) and can just build the expression as "TS + C". It makes the code a lot easier to follow.

* C - TS is not allowed.
* TS + C is allowed.
* C + TS is allowed and equivalent to TS + C (actually rewritten to this form).
* C1 + TS - C2 is allowed and equivalent to (TS + C1) - C2 (actually rewritten to this form).

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql [hunk 5]
```diff
@@ -0,0 +1,11 @@
+-- This file and its contents are licensed under the Timescale License.
+-- Please see the included NOTICE for copyright information and
+-- LICENSE-TIMESCALE for a copy of the license.
+
+-- all of these should have all chunk exclusion happening at plan time
+EXPLAIN SELECT time FROM metrics WHERE time < '2000-01-01'::timestamptz - '6h'::interval;
```
Comment (mkindahl) [inline tsl/test/shared/sql/constify_timestamptz_op_interval.sql:6]: Maybe use `EXPLAIN (VERBOSE ON) SELECT ...`. Could be good for debugging to see the expression that it is being re-written to.

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql (hunk 3 above)
Comment (svenklemm) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: i'm gonna add more tests of course but none of the tests you are suggested are targeted by this PR so i'm not gonna add any of these suggestions

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql (hunk 5 above)
Comment (svenklemm) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:6]: this wouldnt show more than whats already shown but tests are still WIP

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql (hunk 3 above)
Comment (mkindahl) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: Well...

```
mats=# explain verbose select interval '3 days' + timestamptz '2020-06-01 10:11:12';
                                     QUERY PLAN                                      
-------------------------------------------------------------------------------------
 Result  (cost=0.00..0.01 rows=1 width=8)
   Output: ('2020-06-01 10:11:12+02'::timestamp with time zone + '3 days'::interval)
(2 rows)
```

Comment (mkindahl) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: I would assume that your PR will capture the rewritten expression?

Comment (svenklemm) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: TIMESTAMP != TIMESTAMPTZ

Comment (mkindahl) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:11]: Ah, when I wrote TIMESTAMP below, I meant in the generic sense. Please replace with `TIMESTAMPTZ`. :)

File: src/plan_expand_hypertable.c [hunk 6]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+	{
+		op = castNode(OpExpr, left);
+	}
+	else
+		return constraint;
+
+	Oid plus_oid = ts_get_operator("+", PG_CATALOG_NAMESPACE, TIMESTAMPTZOID, INTERVALOID);
+	Oid minus_oid = ts_get_operator("-", PG_CATALOG_NAMESPACE, TIMESTAMPTZOID, INTERVALOID);
+
+	if (op->opresulttype != TIMESTAMPTZOID || (op->opno != plus_oid && op->opno != minus_oid) ||
```
Comment (mkindahl) [inline src/plan_expand_hypertable.c:162]: Is there any specific reason only `TIMESTAMPTZ` is supported and not `TIMESTAMP`?

File: src/plan_expand_hypertable.c (hunk 4 above)
Comment (svenklemm) [reply src/plan_expand_hypertable.c:195]: yes you do because it will be a different operator for other configurations
TIMESTAMPTZ + INTERVAL is different operator from INTERVAL + TIMESTAMPTZ

File: src/plan_expand_hypertable.c (hunk 6 above)
Comment (svenklemm) [reply src/plan_expand_hypertable.c:162]: yes because all the TIMESTAMP operators are marked immutable so they will be constified during planning already

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql (hunk 5 above)
Comment (svenklemm) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql:6]: VERBOSE shows target lists and none of this PR modifies target lists, this is an internal constification to make those expressions available for plan time chunk exclusion, you can only see the effect in chunkappend execution time exclusion not having to do anything

File: src/plan_expand_hypertable.c [hunk 7]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+/*
+ * Constify expressions of the following form in WHERE clause:
+ *
+ * column OP timestamptz - interval
+ * column OP timestamptz + interval
+ *
+ * Iff interval has no month component.
+ *
+ * Since the operators for timestamptz OP interval are marked
+ * as stable they will not be constified during planning a
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:128]:
```suggestion
 * as stable they will not be constified during planning. However, a
```

File: src/plan_expand_hypertable.c [hunk 8]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+ * subset of these operations are immutable and can be safely
+ * constified during planning. We adjust those expressions here
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:130]: Which subset is it? Please clarify.

File: src/plan_expand_hypertable.c [hunk 9]
```diff
@@ -115,6 +116,93 @@ const_datum_get_int(Const *cnst)
...
+		var_on_left = true;
+	}
+	else if (IsA(left, OpExpr) && IsA(right, Var))
+	{
+		op = castNode(OpExpr, left);
+	}
+	else
+		return constraint;
+
+	Oid plus_oid = ts_get_operator("+", PG_CATALOG_NAMESPACE, TIMESTAMPTZOID, INTERVALOID);
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:159]: Nit: mixed declaration and code.

File: src/plan_expand_hypertable.c (hunk 6 above)
Comment (erimatnor) [inline src/plan_expand_hypertable.c:162]: Suggestion: For clarity, I would wrap this condition in an aptly-named function that describes the condition, e.g.,: `is_timestamptz_and_interval_op` or similar. It is a bit hard to follow what this condition is "defining" otherwise.

File: tsl/test/shared/sql/constify_timestamptz_op_interval.sql.in [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+-- test on space-partitioned hypertable
+:PREFIX SELECT time FROM metrics_space WHERE time < '2000-01-01'::timestamptz - '6h'::interval AND device_id = 1;
+-- test on compressed hypertable
+:PREFIX SELECT time FROM metrics_compressed WHERE time < '2000-01-01'::timestamptz - '6h'::interval;
+-- test on space-partitioned compressed hypertable
+:PREFIX SELECT time FROM metrics_space_compressed WHERE time < '2000-01-01'::timestamptz - '6h'::interval AND device_id = 1;
+
+-- month/year intervals are not constified
+:PREFIX SELECT time FROM metrics WHERE time < '2000-02-01'::timestamptz - '1 month'::interval;
+:PREFIX SELECT time FROM metrics WHERE time < '2000-02-01'::timestamptz - '1 month - 1 day'::interval;
```
Comment (mkindahl) [inline tsl/test/shared/sql/constify_timestamptz_op_interval.sql.in:39]: Please add tests for month plus something else as well. If a change is made that for some reason check the non-month part first and accidentally skip checking the month, it would proceed without an error even though it shouldn't. Yes, I'm being paranoid, but I've seen these kind of errors more often than I like.

Comment (svenklemm) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql.in:39]: i'll add but in the interval representation month is stored separately so '1 month - 1 day' will be {.month = 1, .day = -1 } days wont reduce the month value

Comment (mkindahl) [reply tsl/test/shared/sql/constify_timestamptz_op_interval.sql.in:39]: Yes, I've seen the code, but I just want to have tests for both positive and negative "day" parts because it sounds like something that developers can "abuse". E.g., "If it is negative, I know it is a month involved, otherwise it is not so...".

File: src/plan_expand_hypertable.c [hunk 11]
```diff
@@ -93,6 +94,31 @@ is_time_bucket_function(Expr *node)
 	return false;
 }
 
+/*
+ * precheck if an expression is elegible for constification
+ * more thorough check is in constify_timestamptz_op_interval
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:99]:
```suggestion
 * Pre-check to determine if an expression is eligible for constification. 
 * A more thorough check is in constify_timestamptz_op_interval.
```

File: src/plan_expand_hypertable.c [hunk 12]
```diff
@@ -115,6 +141,113 @@ const_datum_get_int(Const *cnst)
...
+ * column OP timestamptz + interval
+ * column OP interval + timestamptz
+ *
+ * Iff interval has no month component.
+ *
+ * Since the operators for timestamptz OP interval are marked
+ * as stable they will not be constified during planning.
+ * However, intervals without a month component can be safely
+ * constified during planning as the result of those calculations
+ * does not depend on timezone settings.
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:157]:
```suggestion
 * do not depend on the timezone setting.
```

File: src/plan_expand_hypertable.c [hunk 13]
```diff
@@ -115,6 +141,113 @@ const_datum_get_int(Const *cnst)
...
+ */
+static OpExpr *
+constify_timestamptz_op_interval(PlannerInfo *root, OpExpr *constraint)
+{
+	Assert(constraint->args->length == 2);
```
Comment (erimatnor) [inline src/plan_expand_hypertable.c:162]: Compilers often complain about having code before declaration. This gets particularly weird with Asserts since they're only available in Debug builds. I would break this up to put the assertion after declarations.
